incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -1983,26 +1946,23 @@ private void onDecodeSubscribe(
                 if (session)
                 {
                     final MqttSessionStateFW.Builder sessionStateBuilder =
-                        mqttSessionStateFW.wrap(sessionStateBuffer, 0, sessionStateBuffer.capacity())
-                            .clientId(clientId);
+                        mqttSessionStateFW.wrap(sessionStateBuffer, 0, sessionStateBuffer.capacity());
 
-                    newSubscriptions.forEach(subscription ->
+                    List<Subscription> currentSubscriptions = new ArrayList<>(sessionStream.subscriptions);
+                    currentSubscriptions.addAll(newSubscriptions);
+                    sessionStream.unAckedSubscriptions.addAll(newSubscriptions);
+                    currentSubscriptions.forEach(subscription ->
+                        sessionStateBuilder.subscriptionsItem(subscriptionBuilder ->
                         {
-                            sessionStateBuilder.subscriptionsItem(subscriptionBuilder ->
-                            {
-                                subscriptionBuilder.subscriptionId(subscription.id);
-                                subscriptionBuilder.flags(subscription.flags);
-                                subscriptionBuilder.pattern(subscription.filter);
-                            });
-                            sessionStream.unAckedSubscriptions.add(subscription);
-                        }
+                            subscriptionBuilder.subscriptionId(subscription.id);
+                            subscriptionBuilder.flags(subscription.flags);
+                            subscriptionBuilder.pattern(subscription.filter);
+                        })
                     );
 
                     final MqttSessionStateFW sessionState = sessionStateBuilder.build();
                     final int payloadSize = sessionState.sizeof();
-
-                    //TODO: is this correct? What is this?
-                    int reserved = payloadSize;
+                    final int reserved = payloadSize + sessionStream.initialPad;
```

Comment:
Since `initialPad` is private to `session` stream, better to handle it there, inside `doSessionData` instead of trying to calculate and pass to `doSessionData` from outside.